From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 20 Jul 2025 22:57:45 +0300
Subject: [PATCH] Jade Protocol


diff --git a/src/main/java/org/purpurmc/purpur/util/MinecraftInternalPlugin.java b/src/main/java/org/purpurmc/purpur/util/MinecraftInternalPlugin.java
index 77a18f3048b18d38c5bd0129065efff27fa774fa..2193a6a9989dcbbd00d23e9d6985d638487a7d60 100644
--- a/src/main/java/org/purpurmc/purpur/util/MinecraftInternalPlugin.java
+++ b/src/main/java/org/purpurmc/purpur/util/MinecraftInternalPlugin.java
@@ -19,6 +19,9 @@ import java.io.InputStream;
 import java.util.List;
 
 public class MinecraftInternalPlugin extends PluginBase {
+
+    public static final MinecraftInternalPlugin INSTANCE = new MinecraftInternalPlugin(); // Horizon - Jade Protocol
+
     private boolean enabled = true;
 
     private final String pluginName;
