From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 18:51:00 +0300
Subject: [PATCH] Add Config


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 45cb1ef06f0daa266e60fdba4f88b5e577dba92f..62aafb8f14d6c5d4ee2aa7294b3402323f43f4c5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1036,6 +1036,7 @@ public final class CraftServer implements Server {
             playerMetadata.removeAll(plugin);
         }
         // Paper end
+        dev.gideonwhite1029.horizon.HorizonConfig.init((File) console.options.valueOf("horizon-settings")); // Horizon - Server Config
         this.reloadData();
         org.spigotmc.SpigotConfig.registerCommands(); // Spigot
         io.papermc.paper.command.PaperCommands.registerCommands(this.console); // Paper
@@ -2812,6 +2813,13 @@ public final class CraftServer implements Server {
         }
         // Purpur end - Purpur config files
 
+        // Horizon start
+        @Override
+        public YamlConfiguration getHorizonConfig() {
+            return dev.gideonwhite1029.horizon.HorizonConfig.config;
+        }
+        // Horizon end
+
         @Override
         public void restart() {
             CraftServer.this.restart();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index e6ea34b18c6db63efa77bb67d3484e4d66315ed6..c4229bc6f69c40fb5d789d0685f94cf3b5978f82 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -147,6 +147,14 @@ public class Main {
                     .describedAs("Yml file");
                 // Purpur end - Purpur config files
 
+                // Horizon start - Server Config
+                acceptsAll(asList("horizon", "horizon-settings"), "File for horizon settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("horizon.yml"))
+                        .describedAs("Yml file");
+                // Horizon end - Server Config
+
                 this.accepts("server-name", "Name of the server")
                         .withRequiredArg()
                         .ofType(String.class)
