From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 18:51:00 +0300
Subject: [PATCH] Add Config


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index e0a84fe6da6b408cbcb2bd783c5fae32c42e0882..e1dd0b6053d295eedf7059068b81d76ebb5c123f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1132,6 +1132,7 @@ public final class CraftServer implements Server {
             playerMetadata.removeAll(plugin);
         }
         // Paper end
+        dev.gideonwhite1029.horizon.HorizonConfig.init((File) console.options.valueOf("horizon-settings")); // Horizon - Server Config
         this.reloadData();
         org.spigotmc.SpigotConfig.registerCommands(); // Spigot
         io.papermc.paper.command.PaperCommands.registerCommands(this.console); // Paper
@@ -3134,6 +3135,13 @@ public final class CraftServer implements Server {
         }
         // Purpur end - Purpur config files
 
+        // Horizon start
+        @Override
+        public YamlConfiguration getHorizonConfig() {
+            return dev.gideonwhite1029.horizon.HorizonConfig.config;
+        }
+        // Horizon end
+
         @Override
         public void restart() {
             org.spigotmc.RestartCommand.restart();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 2e1b7f613de8876095ef39bb0341a3f9520c8d5d..88bb232ce064219e70c648f755a3031d3d095d68 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -183,6 +183,13 @@ public class Main {
                     .defaultsTo(new File("purpur.yml"))
                     .describedAs("Yml file");
                 // Purpur end - Purpur config files
+                // Horizon start - Server Config
+                acceptsAll(asList("horizon", "horizon-settings"), "File for horizon settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("horizon.yml"))
+                        .describedAs("Yml file");
+                // Horizon end - Server Config
                 // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
                         .withRequiredArg()
