From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 18:51:00 +0300
Subject: [PATCH] Add Config


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 49d852817ada69e826b783eb1b246d41ecfe196d..432366151bd56c0ea0cc8719ebf831ffbd587521 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1111,6 +1111,7 @@ public final class CraftServer implements Server {
             playerMetadata.removeAll(plugin);
         }
         // Paper end
+        dev.gideonwhite1029.horizon.HorizonConfig.init((File) console.options.valueOf("horizon-settings")); // Horizon - Server Config
         this.reloadData();
         org.spigotmc.SpigotConfig.registerCommands(); // Spigot
         io.papermc.paper.command.PaperCommands.registerCommands(this.console); // Paper
@@ -3028,6 +3029,13 @@ public final class CraftServer implements Server {
         }
         // Purpur end - Purpur config files
 
+        // Horizon start
+        @Override
+        public YamlConfiguration getHorizonConfig() {
+            return dev.gideonwhite1029.horizon.HorizonConfig.config;
+        }
+        // Horizon end
+
         @Override
         public void restart() {
             CraftServer.this.restart();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 7091dcb86cad81ed4429f11bf5876b4e192a93f5..c5414ce54f2bca766472246267aba11020c39316 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -172,7 +172,14 @@ public class Main {
                     .defaultsTo(new File("purpur.yml"))
                     .describedAs("Yml file");
                 // Purpur end - Purpur config files
-
+                // Horizon start - Server Config
+                acceptsAll(asList("horizon", "horizon-settings"), "File for horizon settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("horizon.yml"))
+                        .describedAs("Yml file");
+                // Horizon end - Server Config
+                // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
                         .withRequiredArg()
                         .ofType(String.class)
