From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Mon, 20 Jan 2025 14:43:43 +0300
Subject: [PATCH] Horizon Protocol Core


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index e1dd0b6053d295eedf7059068b81d76ebb5c123f..d32c85affcff7640f12689eff15c35390c732179 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -507,6 +507,7 @@ public final class CraftServer implements Server {
         this.potionBrewer = new io.papermc.paper.potion.PaperPotionBrewer(console); // Paper - custom potion mixes
         datapackManager = new io.papermc.paper.datapack.PaperDatapackManager(console.getPackRepository()); // Paper
         this.spark = new io.papermc.paper.SparksFly(this); // Paper - spark
+        dev.gideonwhite1029.horizon.protocol.core.HorizonProtocolManager.init(); // Horizon - protocol
     }
 
     public boolean getCommandBlockOverride(String command) {
@@ -1140,6 +1141,7 @@ public final class CraftServer implements Server {
         org.purpurmc.purpur.PurpurConfig.registerCommands(); // Purpur - Purpur config files
         this.overrideAllCommandBlockCommands = this.commandsConfiguration.getStringList("command-block-overrides").contains("*");
         this.ignoreVanillaPermissions = this.commandsConfiguration.getBoolean("ignore-vanilla-permissions");
+        dev.gideonwhite1029.horizon.protocol.core.HorizonProtocolManager.handleServerReload(); // Horizon - protocol
 
         int pollCount = 0;
 
