From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Mon, 20 Jan 2025 15:34:12 +0300
Subject: [PATCH] Disable Packet Limit


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 3f5e76e4df8ff8152060449f89b522910adee834..1cb8091e9f66b6717dcc89e4d63fa181757fc17f 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -286,7 +286,7 @@ public class GlobalConfiguration extends ConfigurationPart {
             }
 
             public boolean isEnabled() {
-                return this.interval > 0.0 && this.maxPacketRate > 0.0;
+                return !dev.gideonwhite1029.horizon.HorizonConfig.disablePacketLimit && (this.interval > 0.0 && this.maxPacketRate > 0.0); // Horizon - disable packet limit
             }
 
             public enum ViolateAction {
