From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Fri, 7 Feb 2025 00:24:13 +0300
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 012fe28a7d8a1f16db11d7bf352874e8ecf6394e..0a9f8b9e2d389156c0ff5c5bf60a348fbe30c6b6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -322,6 +322,7 @@ public final class CraftServer implements Server {
     private final io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler globalRegionScheduler = new io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler();
 
     private final dev.gideonwhite1029.horizon.entity.CraftPhotographerManager photographerManager = new dev.gideonwhite1029.horizon.entity.CraftPhotographerManager(); // Horizon - photographer
+    private final dev.gideonwhite1029.horizon.bytebuf.InternalBytebufHandler internalBytebufHandler = new dev.gideonwhite1029.horizon.bytebuf.InternalBytebufHandler(); // Horizon - Bytebuf API
 
     @Override
     public final io.papermc.paper.threadedregions.scheduler.RegionScheduler getRegionScheduler() {
@@ -3417,4 +3418,15 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Horizon end - replay mod api
+
+    // Horizon start - Bytebuf API
+    @Override
+    public dev.gideonwhite1029.horizon.bytebuf.BytebufManager getBytebufManager() {
+        return internalBytebufHandler.getManager();
+    }
+
+    public dev.gideonwhite1029.horizon.bytebuf.InternalBytebufHandler getBytebufHandler() {
+        return internalBytebufHandler;
+    }
+    // Horizon end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index e317ad68439a26e6e997f8a03df444e786daf1d5..cc87e7da2e6934c63150f395a56b2608d5ce75be 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3581,6 +3581,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end - entity effect API
 
+    // Horizon start - Bytebuf API
+    @Override
+    public void sendPacket(dev.gideonwhite1029.horizon.bytebuf.packet.Packet packet) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), packet);
+    }
+
+    @Override
+    public void sendPacket(dev.gideonwhite1029.horizon.bytebuf.Bytebuf bytebuf, dev.gideonwhite1029.horizon.bytebuf.packet.PacketType type) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), new dev.gideonwhite1029.horizon.bytebuf.packet.Packet(type, bytebuf));
+    }
+    // Horizon end - Bytebuf API
+
     @Override
     public @NotNull PlayerGiveResult give(@NotNull final Collection<@NotNull ItemStack> items, final boolean dropIfFull) {
         Preconditions.checkArgument(items != null, "items cannot be null");
