From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Fri, 7 Feb 2025 00:24:13 +0300
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 97eee300081268c673f0bece2861a54e227f8ed7..4390dfcc4b89a236667517421e0ed4b0ca4e3e58 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -322,6 +322,7 @@ public final class CraftServer implements Server {
     private final io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler globalRegionScheduler = new io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler();
 
     private final dev.gideonwhite1029.horizon.entity.CraftPhotographerManager photographerManager = new dev.gideonwhite1029.horizon.entity.CraftPhotographerManager(); // Horizon - photographer
+    private final dev.gideonwhite1029.horizon.bytebuf.InternalBytebufHandler internalBytebufHandler = new dev.gideonwhite1029.horizon.bytebuf.InternalBytebufHandler(); // Horizon - Bytebuf API
 
     @Override
     public final io.papermc.paper.threadedregions.scheduler.RegionScheduler getRegionScheduler() {
@@ -3411,4 +3412,15 @@ public final class CraftServer implements Server {
         return photographerManager;
     }
     // Horizon end - replay mod api
+
+    // Horizon start - Bytebuf API
+    @Override
+    public dev.gideonwhite1029.horizon.bytebuf.BytebufManager getBytebufManager() {
+        return internalBytebufHandler.getManager();
+    }
+
+    public dev.gideonwhite1029.horizon.bytebuf.InternalBytebufHandler getBytebufHandler() {
+        return internalBytebufHandler;
+    }
+    // Horizon end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 54a267a3f2e959b3c3d38107994f9bda7714ede2..b977fd8b3c42b0be0590dff27be9eb2c566b8604 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3582,6 +3582,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
     // Paper end - entity effect API
 
+    // Horizon start - Bytebuf API
+    @Override
+    public void sendPacket(dev.gideonwhite1029.horizon.bytebuf.packet.Packet packet) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), packet);
+    }
+
+    @Override
+    public void sendPacket(dev.gideonwhite1029.horizon.bytebuf.Bytebuf bytebuf, dev.gideonwhite1029.horizon.bytebuf.packet.PacketType type) {
+        this.server.getBytebufHandler().applyPacketToPlayer(this.getHandle(), new dev.gideonwhite1029.horizon.bytebuf.packet.Packet(type, bytebuf));
+    }
+    // Horizon end - Bytebuf API
+
     @Override
     public @NotNull PlayerGiveResult give(@NotNull final Collection<@NotNull ItemStack> items, final boolean dropIfFull) {
         Preconditions.checkArgument(items != null, "items cannot be null");
