From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Mon, 17 Jun 2024 11:32:00 +0400
Subject: [PATCH] Crafter 1 gt delay

Powered by Leaves: https://github.com/LeavesMC/Leaves

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index 072f998e46b8ee663b668a4fdb5195677f996fdd..e97c75e30dc1d4f5fa71cfbee2dd43087f4d4be6 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -124,4 +124,7 @@ public final class HorizonConfig {
     @GlobalConfig(name = "copper-bulb-1-gt-delay", category = {"features"})
     public static boolean copperBulb1gt = false;
 
+    @GlobalConfig(name = "crafter-1-gt-delay", category = {"features"})
+    public static boolean crafter1gt = false;
+
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
index deb5b4fab94618c08a51e432c33e49b1627bd717..128db0ffe5ebc2bff6fb407a943007ceb80a69e1 100644
--- a/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CrafterBlock.java
@@ -4,6 +4,7 @@ import com.mojang.serialization.MapCodec;
 import java.util.Iterator;
 import java.util.Optional;
 import javax.annotation.Nullable;
+import dev.horizonmc.horizon.HorizonConfig;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -90,7 +91,7 @@ public class CrafterBlock extends BaseEntityBlock {
         BlockEntity tileentity = world.getBlockEntity(pos);
 
         if (flag1 && !flag2) {
-            world.scheduleTick(pos, (Block) this, 4);
+            world.scheduleTick(pos, (Block) this, !HorizonConfig.crafter1gt ? 4 : 1); // Horizon - crafter 1 gt delay
             world.setBlock(pos, (BlockState) state.setValue(CrafterBlock.TRIGGERED, true), 2);
             this.setBlockEntityTriggered(tileentity, true);
         } else if (!flag1 && flag2) {
@@ -156,7 +157,7 @@ public class CrafterBlock extends BaseEntityBlock {
     @Override
     public void setPlacedBy(Level world, BlockPos pos, BlockState state, LivingEntity placer, ItemStack itemStack) {
         if ((Boolean) state.getValue(CrafterBlock.TRIGGERED)) {
-            world.scheduleTick(pos, (Block) this, 4);
+            world.scheduleTick(pos, (Block) this, !HorizonConfig.crafter1gt ? 4 : 1); // Horizon - crafter 1 gt delay
         }
 
     }
