From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Thu, 28 Nov 2024 08:17:13 +0300
Subject: [PATCH] Canvas: Add Eclipse

Powered by Canvas: https://github.com/CraftCanvasMC/Canvas

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index cf1874c4eff8bad7c6fdc98ceaf942eddc0fb47e..b4896c2bf3bf2f4f8fcb1e96c2badab7976f78ec 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -328,4 +328,7 @@ public final class HorizonConfig {
     public static String elytraAeronauticsNoChunkEndMes = "Flight exit cruise mode";
     // Horizon end - elytra-aeronautics
 
+    @GlobalConfig(name = "disable-eclipse", category = {"features"})
+    public static boolean disableEclipse = false;
+
 }
\ No newline at end of file
diff --git a/src/main/java/dev/horizonmc/horizon/plugins/EclipseProviderSource.java b/src/main/java/dev/horizonmc/horizon/plugins/EclipseProviderSource.java
new file mode 100644
index 0000000000000000000000000000000000000000..e5f26aed9c0af68839d4723d61826e09480d600b
--- /dev/null
+++ b/src/main/java/dev/horizonmc/horizon/plugins/EclipseProviderSource.java
@@ -0,0 +1,91 @@
+package dev.horizonmc.horizon.plugins;
+
+import com.mojang.logging.LogUtils;
+import dev.horizonmc.horizon.HorizonConfig;
+import io.papermc.paper.plugin.entrypoint.Entrypoint;
+import io.papermc.paper.plugin.entrypoint.EntrypointHandler;
+import io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler;
+import io.papermc.paper.plugin.provider.PluginProvider;
+import io.papermc.paper.plugin.provider.source.FileProviderSource;
+import io.papermc.paper.plugin.provider.source.ProviderSource;
+import java.io.File;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.bukkit.plugin.java.JavaPlugin;
+import org.slf4j.Logger;
+
+public class EclipseProviderSource implements ProviderSource<Path, Path> {
+
+    public static final EclipseProviderSource INSTANCE = new EclipseProviderSource();
+    public static final String VERSION = "1.3.3";
+    public static final String DOWNLOAD_SOURCE = "https://cdn.modrinth.com/data/2ri1uTe6/versions/UZIxcM1U/eclipse-" + VERSION + "-all.jar";
+    private static final FileProviderSource FILE_PROVIDER_SOURCE = new FileProviderSource("File '%s' specified by Canvas"::formatted);
+    private static final Logger LOGGER = LogUtils.getClassLogger();
+
+    private static boolean hasEclipse() {
+        for (PluginProvider<JavaPlugin> provider : LaunchEntryPointHandler.INSTANCE.get(Entrypoint.PLUGIN).getRegisteredProviders()) {
+            if (provider.getMeta().getName().equalsIgnoreCase("eclipse")) {
+                return true;
+            }
+        }
+        try {
+            Class<?> boot = Class.forName("me.dueris.eclipse.ignite.IgniteBootstrap");
+            return ((AtomicBoolean) boot.getDeclaredField("BOOTED").get(null)).get();
+        } catch (ClassNotFoundException e) {
+            return false;
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException("Unable to reflect into eclipse! Corrupted?", e);
+        } catch (LinkageError ignored) {
+        }
+        return true;
+    }
+
+    @Override
+    public Path prepareContext(final Path context) {
+        // Give server owners opportunity to disable eclipse
+        if (HorizonConfig.disableEclipse || hasEclipse()) {
+            return null;
+        }
+        LOGGER.info("Downloading and bundling Eclipse version '{}'...", VERSION);
+        try {
+            File file = context.toFile();
+            file.getParentFile().mkdirs();
+
+            if (!file.exists()) {
+
+                URLConnection urlConnection = new URL(DOWNLOAD_SOURCE).openConnection();
+                urlConnection.setReadTimeout(5000);
+                urlConnection.setConnectTimeout(5000);
+
+                // Update!
+                LOGGER.info("Updating bundled Eclipse version...");
+                Files.copy(urlConnection.getInputStream(), file.toPath(), StandardCopyOption.REPLACE_EXISTING);
+            }
+
+            // register Eclipse
+            System.setProperty("eclipse.isprovidercontext", "true");
+            return FILE_PROVIDER_SOURCE.prepareContext(context);
+        } catch (Throwable throwable) {
+            LOGGER.error("An unhandled exception occurred when attempting to load the Eclipse provider!", throwable);
+        }
+        return null;
+    }
+
+    @Override
+    public void registerProviders(final EntrypointHandler entrypointHandler, final Path context) throws Exception {
+        if (context == null) {
+            return;
+        }
+
+        try {
+            FILE_PROVIDER_SOURCE.registerProviders(entrypointHandler, context);
+        } catch (IllegalArgumentException ignored) {
+        } catch (Throwable throwable) {
+            LOGGER.error("Error loading Eclipse plugin!: {}", throwable.getMessage(), throwable);
+        }
+    }
+}
diff --git a/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java b/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java
index 70413fddd23ca1165cb5090cce4fddcb1bbca93f..adba4a400a7b4e20619eb021778a4b602afb7b48 100644
--- a/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java
+++ b/src/main/java/io/papermc/paper/plugin/PluginInitializerManager.java
@@ -111,6 +111,7 @@ public class PluginInitializerManager {
 
         // Register the default plugin directory
         io.papermc.paper.plugin.util.EntrypointUtil.registerProvidersFromSource(io.papermc.paper.plugin.provider.source.DirectoryProviderSource.INSTANCE, pluginSystem.pluginDirectoryPath());
+        io.papermc.paper.plugin.util.EntrypointUtil.registerProvidersFromSource(dev.horizonmc.horizon.plugins.EclipseProviderSource.INSTANCE, new File("cache", "eclipse-{}.jar".replace("{}", dev.horizonmc.horizon.plugins.EclipseProviderSource.VERSION)).toPath()); // Horizon - include Eclipse
 
         // Register plugins from the flag
         @SuppressWarnings("unchecked")
