From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 24 Mar 2024 21:27:16 +0400
Subject: [PATCH] Horizon config files


diff --git a/build.gradle.kts b/build.gradle.kts
index ff4b6b735bf0117845cdb316af115079937792ed..1d9b4fe8840784ce4dd23023920d65cff036965a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -161,6 +161,13 @@ tasks.jar {
     }
 }
 
+// Horizon start
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation")
+    options.isWarnings = false
+}
+// Horizon end
+
 tasks.withType<Javadoc> {
     (options as StandardJavadocDocletOptions).addStringOption("-add-modules", "jdk.incubator.vector") // Purpur - our javadocs need this for pufferfish's SIMD patch
     (options as StandardJavadocDocletOptions).addStringOption("Xdoclint:none", "-quiet") // Purpur - silence Paper's bajillion javadoc warnings
@@ -249,3 +256,5 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+
+tasks.getByName("jar").dependsOn("sourcesJar") // Horizon
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index fcfad39173ecf2573a1ba77236bce8d9f73e02bb..52899e58d85cb6cc231fee84a6e500ab1c0f7e98 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2295,6 +2295,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Horizon start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getHorizonConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Horizon end
+
         /**
          * Sends the component to the player
          *
