From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 17:32:40 +0300
Subject: [PATCH] Add Logger


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 8a319e381e93c0ac1c375fd5d3fa965d0413b44b..bcc29ad3281c7d4f26454e086394c737ac2dbeb5 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -353,6 +353,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    private CompoundTag horizonData = new CompoundTag(); // Horizon - Horizon ex data
 
     public void inactiveTick() {
     }
@@ -2555,6 +2556,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 compound.putBoolean("Purpur.FireImmune", immuneToFire);
             }
             // Purpur end - Fire immune API
+            // Horizon start
+            if (compound.contains("Horizon.Data")) {
+                horizonData = compound;
+            }
+            // Horizon end
 
             return compound;
         } catch (Throwable var8) {
@@ -5186,6 +5192,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end - Expose entity id counter
+
+    // Horizon start - horizon ex data
+    public CompoundTag getHorizonData() {
+        return horizonData;
+    }
+    // Horizon end - horizon ex data
+
     // Purpur start - Ridables
     @Nullable
     private Player rider = null;
