From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 18:50:27 +0300
Subject: [PATCH] Add Config


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 398666cb3f0acd0972e9497731fc4eecc723566f..b477c1cf5090f7d98b2c3f7b7f68982652a6c93b 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1101,6 +1101,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper end - rewrite chunk system
         // Paper start - Improved watchdog support - move final shutdown items here
         Util.shutdownExecutors();
+        dev.gideonwhite1029.horizon.HorizonConfig.createWorldSections = false; // Horizon
         try {
             net.minecrell.terminalconsole.TerminalConsoleAppender.close(); // Paper - Use TerminalConsoleAppender
         } catch (final Exception ignored) {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 19a25ee84a4211256a1dbb614db9ebf2ce53182f..c1f04f9e0ec8442656802c700d5ccd450589801d 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -294,6 +294,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         */// Purpur end - Purpur config files // Purpur end - Migrate Setting to reintroduce end void rings
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
 
+        dev.gideonwhite1029.horizon.HorizonConfig.init((java.io.File) options.valueOf("horizon-settings")); // Horizon - Server Config
+
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
         // Paper start - Unix domain socket support
