From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 17:32:40 +0300
Subject: [PATCH] Add Logger


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f55db145661cab3cfc9ffa755792be2c3d80f50a..aa18bec93d82a141fb37690e0e673a5aa0405e4f 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -353,6 +353,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    private CompoundTag horizonData = new CompoundTag(); // Horizon - Horizon ex data
 
     public void inactiveTick() {
     }
@@ -2551,6 +2552,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 compound.putBoolean("Purpur.FireImmune", immuneToFire);
             }
             // Purpur end - Fire immune API
+            // Horizon start
+            if (compound.contains("Horizon.Data")) {
+                horizonData = compound;
+            }
+            // Horizon end
 
             return compound;
         } catch (Throwable var8) {
@@ -5169,6 +5175,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return ((ServerLevel) this.level()).isPositionEntityTicking(this.blockPosition());
     }
     // Paper end - Expose entity id counter
+
+    // Horizon start - horizon ex data
+    public CompoundTag getHorizonData() {
+        return horizonData;
+    }
+    // Horizon end - horizon ex data
+
     // Purpur start - Ridables
     @Nullable
     private Player rider = null;
