From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Mon, 20 Jan 2025 21:39:53 +0300
Subject: [PATCH] Async playerdata Save


diff --git a/net/minecraft/world/level/storage/LevelStorageSource.java b/net/minecraft/world/level/storage/LevelStorageSource.java
index 8104f71c30c1fa46c83acdf0b2e58483df9d89cc..e3a97be2fff62a53e1b471f8dfcdd6f76acbc8c4 100644
--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -515,7 +515,10 @@ public class LevelStorageSource {
             CompoundTag compoundTag = serverConfiguration.createTag(registries, hostPlayerNBT);
             CompoundTag compoundTag1 = new CompoundTag();
             compoundTag1.put("Data", compoundTag);
-            this.saveLevelData(compoundTag1);
+            // Horizon start - Async playerdata save
+            Runnable runnable = () -> this.saveLevelData(compoundTag1);
+            dev.gideonwhite1029.horizon.util.AsyncPlayerDataSaving.saveAsync(runnable);
+            // Horizon end - Async playerdata save
         }
 
         private void saveLevelData(CompoundTag tag) {
@@ -602,7 +605,10 @@ public class LevelStorageSource {
             this.checkLock();
             CompoundTag levelDataTagRaw = LevelStorageSource.readLevelDataTagRaw(this.levelDirectory.dataFile());
             modifier.accept(levelDataTagRaw.getCompoundOrEmpty("Data"));
-            this.saveLevelData(levelDataTagRaw);
+            // Horizon start - Async playerdata save
+            Runnable runnable = () -> this.saveLevelData(levelDataTagRaw);
+            dev.gideonwhite1029.horizon.util.AsyncPlayerDataSaving.saveAsync(runnable);
+            // Horizon end - Async playerdata save
         }
 
         public long makeWorldBackup() throws IOException {
diff --git a/net/minecraft/world/level/storage/PlayerDataStorage.java b/net/minecraft/world/level/storage/PlayerDataStorage.java
index ab9282c04c1996b037567d07f95e2b150bcfcd38..0fbadb66d70dfc6edcb1a456065c429e8884513d 100644
--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -32,6 +32,13 @@ public class PlayerDataStorage {
 
     public void save(Player player) {
         if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
+        // Horizon start - Async playerdata save
+        Runnable runnable = () -> save0(player);
+        dev.gideonwhite1029.horizon.util.AsyncPlayerDataSaving.saveAsync(runnable);
+    }
+
+    private void save0(Player player) {
+        // Horizon end - Async playerdata save
         try {
             CompoundTag compoundTag = player.saveWithoutId(new CompoundTag());
             Path path = this.playerDir.toPath();
