From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Tue, 21 Jan 2025 11:01:17 +0300
Subject: [PATCH] TT20: Lag Compensation


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index f998281c71ab43abe708750f05ed6ae0ed022029..23dd49770994720242cb9683d53e6d6f696ed871 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1561,6 +1561,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         this.server.spark.tickStart(); // Paper - spark
         new com.destroystokyo.paper.event.server.ServerTickStartEvent(this.tickCount+1).callEvent(); // Paper - Server Tick Events
+
+        // Horizon start - Lag compensation
+        if (HorizonConfig.tt20LagCompensation) {
+            dev.gideonwhite1029.horizon.util.optimization.LagCompensation.TPSCalculator.onTick();
+        }
+        // Horizon end - Lag Compensation
+
         this.tickCount++;
         this.tickRateManager.tick();
         this.tickChildren(hasTimeLeft);
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index 43cdc2f8fdfdeb1426e386e0084087779ef62754..ec3715b9af882f732eb857153de860a6bc15345d 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -189,7 +189,13 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public int getTickDelay(LevelReader level) {
-        return level.dimensionType().ultraWarm() ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether; // Purpur - Make lava flow speed configurable
+        // Horizon start - Lag compensation
+        if (dev.gideonwhite1029.horizon.HorizonConfig.tt20LagCompensation && dev.gideonwhite1029.horizon.HorizonConfig.tt20LagCompensationLava) {
+            return level.dimensionType().ultraWarm() ? dev.gideonwhite1029.horizon.util.optimization.LagCompensation.tt20(level.getWorldBorder().world.purpurConfig.lavaSpeedNether, true) : dev.gideonwhite1029.horizon.util.optimization.LagCompensation.tt20(level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether, true); // Horizon
+        } else {
+            return level.dimensionType().ultraWarm() ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether; // Purpur - Make lava flow speed configurable
+        }
+        // Horizon end - Lag compensation
     }
 
     @Override
diff --git a/net/minecraft/world/level/material/WaterFluid.java b/net/minecraft/world/level/material/WaterFluid.java
index b248fe1d66940c05d56fc322df61c52ece72e77f..13ea2cc03ad0f474ce7fd9e693fef5a9e816a1a3 100644
--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -126,7 +126,13 @@ public abstract class WaterFluid extends FlowingFluid {
 
     @Override
     public int getTickDelay(LevelReader level) {
-        return 5;
+        // Horizon start - Lag compensation
+        if (dev.gideonwhite1029.horizon.HorizonConfig.tt20LagCompensation && dev.gideonwhite1029.horizon.HorizonConfig.tt20LagCompensationWater) {
+            return dev.gideonwhite1029.horizon.util.optimization.LagCompensation.tt20(5, true);
+        } else {
+            return 5;
+        }
+        // Horizon end - Lag compensation
     }
 
     @Override
