From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Wed, 26 Feb 2025 22:26:57 +0300
Subject: [PATCH] Syncmatica Protocol


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index afe389697a1ccedcc6dd572470a87783168b481d..0ff7c1e766c5de30651ae5b6e2d13268d669cb0b 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -319,8 +319,11 @@ public class ServerGamePacketListenerImpl
         this.signedMessageDecoder = SignedMessageChain.Decoder.unsigned(player.getUUID(), server::enforceSecureProfile);
         this.chatMessageChain = new FutureChain(server.chatExecutor); // CraftBukkit - async chat
         this.tickEndEvent = new io.papermc.paper.event.packet.ClientTickEndEvent(player.getBukkitEntity()); // Paper - add client tick end event
+        this.exchangeTarget = new dev.gideonwhite1029.horizon.protocol.syncmatica.exchange.ExchangeTarget(this); // Horizon
     }
 
+    public final dev.gideonwhite1029.horizon.protocol.syncmatica.exchange.ExchangeTarget exchangeTarget; // Horizon
+
     // Purpur start - AFK API
     private final com.google.common.cache.LoadingCache<org.bukkit.craftbukkit.entity.CraftPlayer, Boolean> kickPermissionCache = com.google.common.cache.CacheBuilder.newBuilder()
         .maximumSize(1000)
