From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sat, 5 Apr 2025 21:57:12 +0300
Subject: [PATCH] Disable Packet Limit


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index dfb14a238bc4658a8e8a30cebcd10a5651539283..75eb1dd83512a5976c40b474836ab6a4789a1044 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -277,8 +277,8 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
                 if (this.stopReadingPackets) {
                     return;
                 }
-                if (this.allPacketCounts != null ||
-                    io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass())) {
+                if (!dev.gideonwhite1029.horizon.HorizonConfig.disablePacketLimit && (this.allPacketCounts != null || // Horizon - can disable
+                    io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.overrides.containsKey(packet.getClass()))) { // Horizon - can disable
                     long time = System.nanoTime();
                     synchronized (PACKET_LIMIT_LOCK) {
                         if (this.allPacketCounts != null) {
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index deb2542de195618addcfbded4b3c05211936f491..c579fe4fcb65a6feef06834d3b47901c6bb1ab2d 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -847,7 +847,7 @@ public class ServerGamePacketListenerImpl
     public void handleCustomCommandSuggestions(ServerboundCommandSuggestionPacket packet) {
         // PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level()); // Paper - AsyncTabCompleteEvent; run this async
         // CraftBukkit start
-        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.getGameProfile()) && !this.server.isSingleplayerOwner(this.player.getGameProfile())) { // Paper - configurable tab spam limits
+        if (!dev.gideonwhite1029.horizon.HorizonConfig.disablePacketLimit && !this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.getGameProfile()) && !this.server.isSingleplayerOwner(this.player.getGameProfile())) { // Paper - configurable tab spam limits // Horizon - can disable
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - Kick event cause // Paper - add proper async disconnect
             return;
         }
@@ -2089,6 +2089,7 @@ public class ServerGamePacketListenerImpl
     private static int getSpamThreshold() { return io.papermc.paper.configuration.GlobalConfiguration.get().spamLimiter.incomingPacketThreshold; } // Paper - Configurable threshold
 
     private boolean checkLimit(long timestamp) {
+        if (dev.gideonwhite1029.horizon.HorizonConfig.disablePacketLimit) return true; // Horizon - disable
         if (this.lastLimitedPacket != -1 && timestamp - this.lastLimitedPacket < getSpamThreshold() && this.limitedPackets++ >= 8) { // Paper - Configurable threshold; raise packet limit to 8
             return false;
         }
@@ -2624,6 +2625,7 @@ public class ServerGamePacketListenerImpl
 
     // Spigot start - spam exclusions
     private void detectRateSpam(String message) {
+        if (dev.gideonwhite1029.horizon.HorizonConfig.disablePacketLimit) return; // Horizon - disable
         // CraftBukkit start - replaced with thread safe throttle
         if (org.spigotmc.SpigotConfig.enableSpamExclusions) {
             for (String exclude : org.spigotmc.SpigotConfig.spamExclusions) {
@@ -3334,7 +3336,7 @@ public class ServerGamePacketListenerImpl
     @Override
     public void handlePlaceRecipe(ServerboundPlaceRecipePacket packet) {
         // Paper start - auto recipe limit
-        if (!org.bukkit.Bukkit.isPrimaryThread()) {
+        if (!dev.gideonwhite1029.horizon.HorizonConfig.disablePacketLimit && !org.bukkit.Bukkit.isPrimaryThread()) { // Horizon - can disable
             if (!this.recipeSpamPackets.isIncrementAndUnderThreshold()) {
                 this.disconnectAsync(net.minecraft.network.chat.Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause // Paper - add proper async disconnect
                 return;
