From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 24 Mar 2024 21:27:16 +0400
Subject: [PATCH] Horizon config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 84b07a7fd7941b2f75b531646b67a124a0607390..813f36d1e91167a6f3fe3002720bef52c3504cc7 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2277,6 +2277,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Horizon start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getHorizonConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Horizon end
+
         /**
          * Sends the component to the player
          *
