From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 24 Mar 2024 21:27:16 +0400
Subject: [PATCH] Horizon config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ceb81148f5fd17f00d1a0fb2df1f31c45aa6807d..85e2dc4fbcb37447238b426c4641c7685a7a4f95 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2263,6 +2263,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Horizon start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getHorizonConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Horizon end
+
         /**
          * Sends the component to the player
          *
