From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Thu, 28 Mar 2024 16:57:54 +0400
Subject: [PATCH] Information changed during startup


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 7196594e07af19a14c320d77df893978525fe386..b770c4dcbccba80e493207b01dec2eddcf0e4172 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -33,6 +33,13 @@ public interface ServerBuildInfo {
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
 
+    // Horizon start
+    /**
+     * The brand id for Horizon.
+     */
+    Key BRAND_HORIZON_ID = Key.key("gideonwhite1029", "horizon");
+    // Horizon end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 3a71f8fe73a2309f57cb06dedf204d3e37b46b03..2b2f4a389699bc22cadaef3d189a036a610aa8f8 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -120,8 +120,18 @@ public final class Bukkit {
       */
     @NotNull
     public static String getVersionMessage() {
-        final io.papermc.paper.ServerBuildInfo version = io.papermc.paper.ServerBuildInfo.buildInfo();
-        return "This server is running " + getName() + " version " + version.asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_FULL) + " (Implementing API version " + getBukkitVersion() + ")";
+        // Horizon start
+        return "\n" +
+                "\u001B[38;5;99m      __ ______  ___  ________  ____  _  __\n" +
+                "\u001B[38;5;135m     / // / __ \\/ _ \\/  _/_  / / __ \\/ |/ /\n" +
+                "\u001B[38;5;171m    / _  / /_/ / , _// /  / /_/ /_/ /    / \n" +
+                "\u001B[38;5;207m   /_//_/\\____/_/|_/___/ /___/\\____/_/|_/  \u001B[0m\n" +
+                "\n" +
+                "  \u001b[38;5;105m Server name: \u001B[0m" + getName() + "\n"+
+                "  \u001b[38;5;105m Version: \u001B[0m" + getVersion() + "\n" +
+                "  \u001b[38;5;105m API version: \u001B[0m" + getBukkitVersion() + "\n";
+
+        // Horizon stop
         // Paper end
     }
 
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index c880d0010849ab733ad13bbd18fab3c864d0cf61..3d5faae011ebab29ca38175df2821e8e993cd10a 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -259,7 +259,7 @@ public class VersionCommand extends BukkitCommand {
         // Purpur start
         int distance = getVersionFetcher().distance();
         final Component message = Component.join(net.kyori.adventure.text.JoinConfiguration.separator(Component.newline()),
-            ChatColor.parseMM("<grey>Current Purpur Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()),
+            ChatColor.parseMM("<grey>Current Horizon Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()), // Horizon
             // Purpur end
             msg
         );
