From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Thu, 28 Mar 2024 16:57:54 +0400
Subject: [PATCH] Information changed during startup


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index afe47ecf0e7ae673aaf62b226e9cb07d2ed393eb..9acad5d8a8bcbcd75057bfa31f50a66eca9caaa9 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -125,7 +125,18 @@ public final class Bukkit {
         if ("master".equals(gitBranch) || "main".equals(gitBranch)) {
             branchMsg = "";  // Don't show branch on main/master
         }
-        return "This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ") (Git: " + gitCommit + branchMsg + ")";
+        // Horizon start
+        return "\n" +
+                "      __ ______  ___  ________  ____  _  __\n" +
+                "     / // / __ \\/ _ \\/  _/_  / / __ \\/ |/ /\n" +
+                "    / _  / /_/ / , _// /  / /_/ /_/ /    / \n" +
+                "   /_//_/\\____/_/|_/___/ /___/\\____/_/|_/  \n" +
+                "\n" +
+                "   Server name: " + getName() + "\n"+
+                "   Version: " + getVersion() + "\n" +
+                "   API version: " + getBukkitVersion() + "\n" +
+                "   Github commit: " + gitCommit + "\n";
+        // Horizon stop
         // Paper end
     }
 
