From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Thu, 28 Mar 2024 16:57:54 +0400
Subject: [PATCH] Information changed during startup


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 70ef8f63ab79e102cb4326c21cc344488f4fbdd3..fd3b2746f4524be1c8da71a953961026f0c2810d 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -127,7 +127,19 @@ public final class Bukkit {
         if ("master".equals(gitBranch) || "main".equals(gitBranch)) {
             branchMsg = "";  // Don't show branch on main/master
         }
-        return "This server is running " + getName() + " version " + getVersion() + " (Implementing API version " + getBukkitVersion() + ") (Git: " + gitCommit + branchMsg + ")";
+        // Horizon start
+        return "\n" +
+                "\u001B[38;5;99m      __ ______  ___  ________  ____  _  __\n" +
+                "\u001B[38;5;135m     / // / __ \\/ _ \\/  _/_  / / __ \\/ |/ /\n" +
+                "\u001B[38;5;171m    / _  / /_/ / , _// /  / /_/ /_/ /    / \n" +
+                "\u001B[38;5;207m   /_//_/\\____/_/|_/___/ /___/\\____/_/|_/  \u001B[0m\n" +
+                "\n" +
+                "  \u001b[38;5;105m Server name: \u001B[0m" + getName() + "\n"+
+                "  \u001b[38;5;105m Version: \u001B[0m" + getVersion() + "\n" +
+                "  \u001b[38;5;105m API version: \u001B[0m" + getBukkitVersion() + "\n" +
+                "  \u001b[38;5;105m Github commit: \u001B[0m" + gitCommit + "\n";
+
+        // Horizon stop
         // Paper end
     }
 
