From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Thu, 28 Mar 2024 16:57:54 +0400
Subject: [PATCH] Information changed during startup


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 75fe3bdbd44440c7cbea45dbd248e260f322dbb2..bf161f9bb61b381a464debb8ca8e389fe294b418 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -119,8 +119,18 @@ public final class Bukkit {
       */
     @NotNull
     public static String getVersionMessage() {
-        final io.papermc.paper.ServerBuildInfo version = io.papermc.paper.ServerBuildInfo.buildInfo();
-        return "This server is running " + getName() + " version " + version.asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_FULL) + " (Implementing API version " + getBukkitVersion() + ")";
+        // Horizon start
+        return "\n" +
+                "\u001B[38;5;99m      __ ______  ___  ________  ____  _  __\n" +
+                "\u001B[38;5;135m     / // / __ \\/ _ \\/  _/_  / / __ \\/ |/ /\n" +
+                "\u001B[38;5;171m    / _  / /_/ / , _// /  / /_/ /_/ /    / \n" +
+                "\u001B[38;5;207m   /_//_/\\____/_/|_/___/ /___/\\____/_/|_/  \u001B[0m\n" +
+                "\n" +
+                "  \u001b[38;5;105m Server name: \u001B[0m" + getName() + "\n"+
+                "  \u001b[38;5;105m Version: \u001B[0m" + getVersion() + "\n" +
+                "  \u001b[38;5;105m API version: \u001B[0m" + getBukkitVersion() + "\n";
+
+        // Horizon stop
         // Paper end
     }
 
