From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Thu, 28 Mar 2024 16:57:54 +0400
Subject: [PATCH] Information changed during startup


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 7107eabf02d9d9b69d97c978ba421ad52031ec3f..f4dbe9daf6521a20a0aa57a110836fb976cffffb 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +25,13 @@ public interface ServerBuildInfo {
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
 
+    // Horizon start
+    /**
+     * The brand id for Horizon.
+     */
+    Key BRAND_HORIZON_ID = Key.key("gideonwhite1029", "horizon");
+    // Horizon end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index bb09c468ab1261ea4eaf59012d30a2e45199caf7..fe0ecdae1b4e09c48ad09806bdd9676b9d77b219 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -119,8 +119,18 @@ public final class Bukkit {
       */
     @NotNull
     public static String getVersionMessage() {
-        final io.papermc.paper.ServerBuildInfo version = io.papermc.paper.ServerBuildInfo.buildInfo();
-        return "This server is running " + getName() + " version " + version.asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_FULL) + " (Implementing API version " + getBukkitVersion() + ")";
+        // Horizon start
+        return "\n" +
+                "\u001B[38;5;99m      __ ______  ___  ________  ____  _  __\n" +
+                "\u001B[38;5;135m     / // / __ \\/ _ \\/  _/_  / / __ \\/ |/ /\n" +
+                "\u001B[38;5;171m    / _  / /_/ / , _// /  / /_/ /_/ /    / \n" +
+                "\u001B[38;5;207m   /_//_/\\____/_/|_/___/ /___/\\____/_/|_/  \u001B[0m\n" +
+                "\n" +
+                "  \u001b[38;5;105m Server name: \u001B[0m" + getName() + "\n"+
+                "  \u001b[38;5;105m Version: \u001B[0m" + getVersion() + "\n" +
+                "  \u001b[38;5;105m API version: \u001B[0m" + getBukkitVersion() + "\n";
+
+        // Horizon stop
         // Paper end
     }
 
