From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sat, 27 Apr 2024 10:17:24 +0400
Subject: [PATCH] Fix Annotations


diff --git a/src/main/java/dev/horizonmc/horizon/entity/BotManager.java b/src/main/java/dev/horizonmc/horizon/entity/BotManager.java
index e1eceeb6d53062885993bcbc0849c46afe878d0f..bbbf17393da94db4a6cfff822a06c8abd0fd0331 100644
--- a/src/main/java/dev/horizonmc/horizon/entity/BotManager.java
+++ b/src/main/java/dev/horizonmc/horizon/entity/BotManager.java
@@ -86,6 +86,7 @@ public interface BotManager {
      *
      * @return a view of fakeplayers.
      */
+    @Nullable
     public Collection<Bot> getBots();
 
     /**
@@ -95,7 +96,7 @@ public interface BotManager {
      * @param action action executor
      * @return true if success, or false
      */
-    public boolean registerCustomBotAction(String name, CustomBotAction action);
+    public boolean registerCustomBotAction(@NotNull String name, @NotNull CustomBotAction action);
 
     /**
      * Unregister a custom bot action.
@@ -103,5 +104,5 @@ public interface BotManager {
      * @param name action name
      * @return true if success, or false
      */
-    public boolean unregisterCustomBotAction(String name);
+    public boolean unregisterCustomBotAction(@NotNull String name);
 }
diff --git a/src/main/java/dev/horizonmc/horizon/entity/PhotographerManager.java b/src/main/java/dev/horizonmc/horizon/entity/PhotographerManager.java
index 1bbf332d0ec975105379fd99fdb7e9f49e359c50..e75e2940a46c89d26963af67c5af0c207da8fc40 100644
--- a/src/main/java/dev/horizonmc/horizon/entity/PhotographerManager.java
+++ b/src/main/java/dev/horizonmc/horizon/entity/PhotographerManager.java
@@ -29,5 +29,6 @@ public interface PhotographerManager {
 
     public void removeAllPhotographers();
 
+    @Nullable
     public Collection<Photographer> getPhotographers();
 }
diff --git a/src/main/java/dev/horizonmc/horizon/entity/botaction/CustomBotAction.java b/src/main/java/dev/horizonmc/horizon/entity/botaction/CustomBotAction.java
index aadc6f81af946e58a94da59415f19c340ae60bcf..aaea1aa5cfbab118034c9da28d42560569644fc2 100644
--- a/src/main/java/dev/horizonmc/horizon/entity/botaction/CustomBotAction.java
+++ b/src/main/java/dev/horizonmc/horizon/entity/botaction/CustomBotAction.java
@@ -18,7 +18,7 @@ public interface CustomBotAction {
      * @param bot bot of the action
      * @return true if once action finish, otherwise false
      */
-    public boolean doTick(Bot bot);
+    public boolean doTick(@NotNull Bot bot);
 
     /**
      * Created a new action instance.
@@ -27,7 +27,7 @@ public interface CustomBotAction {
      * @param args   passed action arguments
      * @return a new action instance with given args
      */
-    public @Nullable CustomBotAction getNew(Player player, String[] args);
+    public @Nullable CustomBotAction getNew(@NotNull Player player, @NotNull String[] args);
 
     /**
      * Requests a list of possible completions for a action argument.
diff --git a/src/main/java/dev/horizonmc/horizon/entity/botaction/HorizonBotAction.java b/src/main/java/dev/horizonmc/horizon/entity/botaction/HorizonBotAction.java
index c9affd016706e645b8933291d2cf7690f8006fb3..a1024619a70eacb8766499f21fd94151dbbe6134 100644
--- a/src/main/java/dev/horizonmc/horizon/entity/botaction/HorizonBotAction.java
+++ b/src/main/java/dev/horizonmc/horizon/entity/botaction/HorizonBotAction.java
@@ -1,5 +1,7 @@
 package dev.horizonmc.horizon.entity.botaction;
 
+import org.jetbrains.annotations.NotNull;
+
 /**
  * A Horizon bot action enum
  */
@@ -26,6 +28,7 @@ public enum HorizonBotAction {
         this.name = name;
     }
 
+    @NotNull
     public String getName() {
         return name;
     }
diff --git a/src/main/java/dev/horizonmc/horizon/event/bot/BotActionEvent.java b/src/main/java/dev/horizonmc/horizon/event/bot/BotActionEvent.java
index dc1ce98326617468f4f2b84e6ddcef51525e86af..fa4b0d31bd0a00a46e1bcb5d2f7dbce2463c9e4e 100644
--- a/src/main/java/dev/horizonmc/horizon/event/bot/BotActionEvent.java
+++ b/src/main/java/dev/horizonmc/horizon/event/bot/BotActionEvent.java
@@ -12,7 +12,7 @@ public class BotActionEvent extends BotEvent implements Cancellable {
     private final String[] actionArgs;
     private boolean cancel = false;
 
-    public BotActionEvent(@NotNull Bot who, String actionName, String[] actionArgs) {
+    public BotActionEvent(@NotNull Bot who, @NotNull String actionName, @NotNull String[] actionArgs) {
         super(who);
         this.actionArgs = actionArgs;
         this.actionName = actionName;
@@ -43,6 +43,7 @@ public class BotActionEvent extends BotEvent implements Cancellable {
         this.cancel = cancel;
     }
 
+    @NotNull
     public static HandlerList getHandlerList() {
         return handlers;
     }
diff --git a/src/main/java/dev/horizonmc/horizon/event/bot/BotConfigModifyEvent.java b/src/main/java/dev/horizonmc/horizon/event/bot/BotConfigModifyEvent.java
index 092794ce8bccfeddbd03768c5e7b74c98afce7ac..0768e731e7a57962fbc6d78a07c3641f8e18e296 100644
--- a/src/main/java/dev/horizonmc/horizon/event/bot/BotConfigModifyEvent.java
+++ b/src/main/java/dev/horizonmc/horizon/event/bot/BotConfigModifyEvent.java
@@ -12,7 +12,7 @@ public class BotConfigModifyEvent extends BotEvent implements Cancellable {
     private final String configValue;
     private boolean cancel;
 
-    public BotConfigModifyEvent(@NotNull Bot who, String configName, String configValue) {
+    public BotConfigModifyEvent(@NotNull Bot who, @NotNull String configName, @NotNull String configValue) {
         super(who);
         this.configName = configName;
         this.configValue = configValue;
@@ -43,6 +43,7 @@ public class BotConfigModifyEvent extends BotEvent implements Cancellable {
         this.cancel = cancel;
     }
 
+    @NotNull
     public static HandlerList getHandlerList() {
         return handlers;
     }
diff --git a/src/main/java/dev/horizonmc/horizon/event/bot/BotCreateEvent.java b/src/main/java/dev/horizonmc/horizon/event/bot/BotCreateEvent.java
index 0610e6bff0be78102cdcbf8478a4c648b997c7c4..f587b1cb16b689dd9975c7f35f50a27b543e86e1 100644
--- a/src/main/java/dev/horizonmc/horizon/event/bot/BotCreateEvent.java
+++ b/src/main/java/dev/horizonmc/horizon/event/bot/BotCreateEvent.java
@@ -31,6 +31,7 @@ public class BotCreateEvent extends Event implements Cancellable {
      *
      * @return fakeplayer name
      */
+    @NotNull
     public String getBot() {
         return bot;
     }
