From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 13:52:38 +0300
Subject: [PATCH] Horizon config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 69c0ca1b0c0cdbc911ba3f053fafc5624539917d..7550b9f93c908fe95a157e2c445b320fcb4d3a57 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2399,6 +2399,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Horizon start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getHorizonConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Horizon end
+
         /**
          * Sends the component to the player
          *
