From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 19 Jan 2025 14:02:35 +0300
Subject: [PATCH] Information changed during startup


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index fb1fe2651e53a9bf46b3632c638e13eea9dcda93..54c4a2e3e7739300b6f2aa6889e80ee976afd038 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +25,14 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
+
+    // Horizon start
+    /**
+     * The brand id for Horizon.
+     */
+    Key BRAND_HORIZON_ID = Key.key("gideonwhite1029", "horizon");
+    // Horizon end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 1c735d025c2d0fd75e850b44e721147dd69e8d34..261f3da54e4deb410461bcf62b1d5454067a440e 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -124,8 +124,17 @@ public final class Bukkit {
     @NotNull
     public static String getVersionMessage() {
         final io.papermc.paper.ServerBuildInfo version = io.papermc.paper.ServerBuildInfo.buildInfo();
-        return "This server is running " + getName() + " version " + version.asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_FULL) + " (Implementing API version " + getBukkitVersion() + ")";
-        // Paper end
+        // Horizon start
+        return "\n" +
+                "\u001B[38;5;99m      __ ______  ___  ________  ____  _  __\n" +
+                "\u001B[38;5;135m     / // / __ \\/ _ \\/  _/_  / / __ \\/ |/ /\n" +
+                "\u001B[38;5;171m    / _  / /_/ / , _// /  / /_/ /_/ /    / \n" +
+                "\u001B[38;5;207m   /_//_/\\____/_/|_/___/ /___/\\____/_/|_/  \u001B[0m\n" +
+                "\n" +
+                "  \u001b[38;5;105m Server name: \u001B[0m" + getName() + "\n"+
+                "  \u001b[38;5;105m Version: \u001B[0m" + getVersion() + "\n" +
+                "  \u001b[38;5;105m API version: \u001B[0m" + getBukkitVersion() + "\n";
+        // Horizon stop
     }
 
     /**
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 424b01d6fc013148f19b039ca7ade9ff484796ee..578026a18453998faefd2f3a6ddb9075fe06b1fc 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -258,7 +258,7 @@ public class VersionCommand extends BukkitCommand {
         // Purpur start
         int distance = getVersionFetcher().distance();
         final Component message = Component.join(net.kyori.adventure.text.JoinConfiguration.separator(Component.newline()),
-            ChatColor.parseMM("<grey>Current Purpur Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()),
+            ChatColor.parseMM("<grey>Current Horizon Version: %s%s*", distance == 0 ? "<green>" : distance > 0 ? "<yellow>" : "<red>", Bukkit.getVersion()), // Horizon
             // Purpur end
             msg
         );
