From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Wed, 5 Feb 2025 20:41:40 +0300
Subject: [PATCH] Replay API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 0a1c3f8cafb1f8f3b8bc48d1dae6cc4745b79234..4252fe01ba88cf6d42b00b96717e13cc1a078685 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3134,4 +3134,20 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    public static @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Horizon end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ef933ffef65e8530d8d540a3faabf896c9cdb00e..caf0441e7afaa6862e6c98f2f4d2570104723aa5 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2824,4 +2824,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager();
+    // Horizon end - Photographer API
 }
