From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Wed, 5 Feb 2025 20:41:40 +0300
Subject: [PATCH] Replay API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 7e5926a5d7329ede564c8e13b7421473d95d6f80..7a8bc0715cd2c86c7448917c2f9a5ae1884f5fa5 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3122,4 +3122,20 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    public static @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Horizon end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 4cf3f23f303374610737f29d75d488e6c10bffad..4282676f170942332b0d96410a30a14bf6480c24 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2784,4 +2784,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager();
+    // Horizon end - Photographer API
 }
