From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Wed, 5 Feb 2025 20:41:40 +0300
Subject: [PATCH] Replay API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 8014d901a4a9fadab7dae878d9c15c0060fd95cc..69bd4dc366bdc73580da413f258ea89d45a2c9b3 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3135,4 +3135,20 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    public static @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Horizon end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 338337723222247d6633674cbb928bd92fd1e8af..ddc7493db1600164472b6922b9f1f795ddba80c3 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2831,4 +2831,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager();
+    // Horizon end - Photographer API
 }
