From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Wed, 5 Feb 2025 20:41:40 +0300
Subject: [PATCH] Replay API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 261f3da54e4deb410461bcf62b1d5454067a440e..054bda6b8b86f413794889cecf2d80b7528db169 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3145,4 +3145,20 @@ public final class Bukkit {
         server.clearBlockHighlights();
     }
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    public static @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager() {
+        return server.getPhotographerManager();
+    }
+    // Horizon end - Photographer API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 3153355d2ba78466d38b108461dbe038c9f9b1b6..aaef3f8f75f051fd3fa41ba2491c26e62e556611 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2842,4 +2842,18 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void clearBlockHighlights();
     // Purpur end - Debug Marker API
+
+    // Horizon start - Photographer API
+    /**
+     * Gets the PhotographerManager instance.
+     * <p>
+     * This method returns the PhotographerManager instance from the server.
+     * The PhotographerManager is responsible for managing photography-related functionalities
+     * within the Horizon replay system.
+     * </p>
+     *
+     * @return the PhotographerManager instance
+     */
+    @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager();
+    // Horizon end - Photographer API
 }
