From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Fri, 7 Feb 2025 00:23:57 +0300
Subject: [PATCH] Bytebuf API


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 69bd4dc366bdc73580da413f258ea89d45a2c9b3..a42972c2ee1e166e1c314431e5f4cad1c910ad97 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -3151,4 +3151,19 @@ public final class Bukkit {
         return server.getPhotographerManager();
     }
     // Horizon end - Photographer API
+    // Horizon start - Bytebuf API
+    /**
+     * Gets the BytebufManager instance.
+     * <p>
+     * This method returns the BytebufManager instance from the server.
+     * The BytebufManager is responsible for managing byte buffer-related functionalities
+     * within the Horizon system.
+     * </p>
+     *
+     * @return the BytebufManager instance
+     */
+    public static dev.gideonwhite1029.horizon.bytebuf.BytebufManager getBytebufManager() {
+        return server.getBytebufManager();
+    }
+    // Horizon end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index aaad621eda0d68e6578a6e356b1f99d692891dd4..bc36666a6f0bda5a80ad41991e7b2adc158e2712 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2844,4 +2844,17 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull dev.gideonwhite1029.horizon.replay.PhotographerManager getPhotographerManager();
     // Horizon end - Photographer API
+    // Horizon start - Bytebuf API
+    /**
+     * Gets the BytebufManager instance.
+     * <p>
+     * This method returns the BytebufManager instance from the server.
+     * The BytebufManager is responsible for managing byte buffer-related functionalities
+     * within the Horizon system.
+     * </p>
+     *
+     * @return the BytebufManager instance
+     */
+    dev.gideonwhite1029.horizon.bytebuf.BytebufManager getBytebufManager();
+    // Horizon end - Bytebuf API
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 7262543459d43006f8758e6623c64a12c10c974c..1b97f044266f3bac2bd1a21a715f70fd1814ec44 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3845,6 +3845,23 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     boolean isChunkSent(long chunkKey);
     // Paper end
 
+    // Horizon start - Bytebuf API
+    /**
+     * Sends a packet to the player.
+     *
+     * @param packet the packet to send
+     */
+    void sendPacket(dev.gideonwhite1029.horizon.bytebuf.packet.Packet packet);
+
+    /**
+     * Sends a packet to the player using a Bytebuf and PacketType.
+     *
+     * @param buf the Bytebuf containing the packet data
+     * @param type the type of the packet
+     */
+    void sendPacket(dev.gideonwhite1029.horizon.bytebuf.Bytebuf buf, dev.gideonwhite1029.horizon.bytebuf.packet.PacketType type);
+    // Horizon end - Bytebuf API
+
     @Override
     Spigot spigot();
     // Spigot end
