From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Mon, 25 Mar 2024 14:26:47 +0400
Subject: [PATCH] Add Logger

Powered by Leaves: https://github.com/LeavesMC/Leaves

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonLogger.java b/src/main/java/dev/horizonmc/horizon/HorizonLogger.java
new file mode 100644
index 0000000000000000000000000000000000000000..1cb0198de00244f30bb09d1398ef3acd41f02a82
--- /dev/null
+++ b/src/main/java/dev/horizonmc/horizon/HorizonLogger.java
@@ -0,0 +1,16 @@
+package dev.horizonmc.horizon;
+
+import org.bukkit.Bukkit;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class HorizonLogger extends Logger {
+    public static final HorizonLogger LOGGER = new HorizonLogger();
+
+    private HorizonLogger() {
+        super("Horizon", null);
+        setParent(Bukkit.getLogger());
+        setLevel(Level.ALL);
+    }
+}
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index d6d992e4bc756acf649dba11c2fc16064ea3d5bc..c181f9c2a35566c6fab39cd704a47dd93a91bf7c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -430,6 +430,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     private UUID originWorld;
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean fixedPose = false; // Paper - Expand Pose API
+    private CompoundTag horizonData = new CompoundTag(); // Horizon - Horizon ex data
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
@@ -2481,6 +2482,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 nbttagcompound.putBoolean("Purpur.FireImmune", immuneToFire);
             }
             // Purpur end
+            nbttagcompound.put("Horizon.Data", horizonData); // Horizon - Horizon ex data
             return nbttagcompound;
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Saving entity NBT");
@@ -2633,6 +2635,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 immuneToFire = nbt.getBoolean("Purpur.FireImmune");
             }
             // Purpur end
+            // Horizon start
+            if (nbt.contains("Horizon.Data")) {
+                horizonData = nbt.getCompound("Horizon.Data");
+            }
+            // Horizon end
 
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Loading entity NBT");
