From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Tue, 26 Mar 2024 10:23:53 +0400
Subject: [PATCH] No Chat Reports


diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index 8fa59b359f80e5f1c805c7fc65ecdcff713b122b..39f2bc18ccf2a942064dfb976d3cb0e3d6d31049 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -80,5 +80,4 @@ public final class HorizonConfig {
     }
     @GlobalConfig(name = "no-chat-reports", category = {"features"})
     public static boolean noChatReports = false;
-
 }
