From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Wed, 27 Mar 2024 12:50:01 +0400
Subject: [PATCH] Spectator dont get Advancement


diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index c27225cc1a2eb9aa983c6ef8afe26aca3567bd94..c5c5604b8ac1735a893a5e5b40a22eb9e367a658 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -94,4 +94,7 @@ public final class HorizonConfig {
     @GlobalConfig(name = "vanilla-hopper", category = {"features"})
     public static boolean vanillaHopper = false;
 
+    @GlobalConfig(name = "spectator-dont-get-advancement", category = {"features"})
+    public static boolean spectatorDontGetAdvancement = false;
+
 }
diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java
index ba8a8575af92541cef2e116743d51cd68d1e794a..32f7c86698ac34f67e488cd5ea2d3f0986896076 100644
--- a/src/main/java/net/minecraft/server/PlayerAdvancements.java
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java
@@ -25,6 +25,8 @@ import java.util.Map.Entry;
 import java.util.Set;
 import java.util.function.BiConsumer;
 import javax.annotation.Nullable;
+
+import dev.horizonmc.horizon.HorizonConfig;
 import net.minecraft.FileUtil;
 import net.minecraft.Util;
 import net.minecraft.advancements.Advancement;
@@ -45,6 +47,7 @@ import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.players.PlayerList;
 import net.minecraft.util.datafix.DataFixTypes;
 import net.minecraft.world.level.GameRules;
+import net.minecraft.world.level.GameType;
 import org.slf4j.Logger;
 
 public class PlayerAdvancements {
@@ -222,6 +225,11 @@ public class PlayerAdvancements {
     }
 
     public boolean award(AdvancementHolder advancement, String criterionName) {
+        // Leaves start - spectator don't get advancement
+        if (HorizonConfig.spectatorDontGetAdvancement && player.gameMode.getGameModeForPlayer() == GameType.SPECTATOR) {
+            return false;
+        }
+        // Leaves end - spectator don't get advancement
         boolean flag = false;
         AdvancementProgress advancementprogress = this.getOrStartProgress(advancement);
         boolean flag1 = advancementprogress.isDone();
