From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sat, 1 Jun 2024 16:10:23 +0400
Subject: [PATCH] No TNT place update

Powered by Leaves: https://github.com/LeavesMC/Leaves

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index 8e71e7a89bf51dbc3153f14c777dfe0b2a21ff53..64d88eca763b26e0c230c5b2c1836668b914c5d1 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -106,4 +106,7 @@ public final class HorizonConfig {
     @GlobalConfig(name = "use_virtual_thread_for_async_scheduler", category = {"utils"})
     public static boolean useVirtualThreadForAsyncScheduler = false;
 
+    @GlobalConfig(name = "no-tnt-place-update", category = {"features"})
+    public static boolean noTNTPlaceUpdate = false;
+
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/level/block/TntBlock.java b/src/main/java/net/minecraft/world/level/block/TntBlock.java
index a3525ae6d83591664e1456f20d9732a8de0ec326..17fcb89c65dd487a4bd7ac3062f2ea889f1244e6 100644
--- a/src/main/java/net/minecraft/world/level/block/TntBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TntBlock.java
@@ -49,7 +49,7 @@ public class TntBlock extends Block {
     @Override
     protected void onPlace(BlockState state, Level world, BlockPos pos, BlockState oldState, boolean notify) {
         if (!oldState.is(state.getBlock())) {
-            if (world.hasNeighborSignal(pos) && CraftEventFactory.callTNTPrimeEvent(world, pos, PrimeCause.REDSTONE, null, null)) { // CraftBukkit - TNTPrimeEvent
+            if (!dev.horizonmc.horizon.HorizonConfig.noTNTPlaceUpdate && world.hasNeighborSignal(pos) && CraftEventFactory.callTNTPrimeEvent(world, pos, PrimeCause.REDSTONE, null, null)) { // CraftBukkit - TNTPrimeEvent // Horizon - No tnt place update
                 // Paper start - TNTPrimeEvent
                 org.bukkit.block.Block tntBlock = org.bukkit.craftbukkit.block.CraftBlock.at(world, pos);
                 if (!new com.destroystokyo.paper.event.block.TNTPrimeEvent(tntBlock, com.destroystokyo.paper.event.block.TNTPrimeEvent.PrimeReason.REDSTONE, null).callEvent()) {
