From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sun, 2 Jun 2024 16:34:50 +0400
Subject: [PATCH] InstantBlockUpdater Reintroduced

Powered by Leaves: https://github.com/LeavesMC/Leaves

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index 64d88eca763b26e0c230c5b2c1836668b914c5d1..a2b16fdce1a86377fdd3e4b29f31e1578b9cf748 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -109,4 +109,7 @@ public final class HorizonConfig {
     @GlobalConfig(name = "no-tnt-place-update", category = {"features"})
     public static boolean noTNTPlaceUpdate = false;
 
+    @GlobalConfig(name = "instant-block-updater-reintroduced", category = {"features"})
+    public static boolean instantBlockUpdaterReintroduced = false;
+
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index eef03486c1f650643cb64fa53a03b78d9a3251d7..fe64e03cc3b52a693a8c4aa79d707e962eb3d3d1 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -74,6 +74,7 @@ import net.minecraft.world.level.lighting.LevelLightEngine;
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.level.redstone.CollectingNeighborUpdater;
+import net.minecraft.world.level.redstone.InstantNeighborUpdater;
 import net.minecraft.world.level.redstone.NeighborUpdater;
 import net.minecraft.world.level.saveddata.maps.MapId;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
@@ -769,7 +770,13 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         this.thread = Thread.currentThread();
         this.biomeManager = new BiomeManager(this, i);
         this.isDebug = flag1;
-        this.neighborUpdater = new CollectingNeighborUpdater(this, j);
+        // Horizon start - instantBlockUpdaterReintroduced
+        if (dev.horizonmc.horizon.HorizonConfig.instantBlockUpdaterReintroduced) {
+            this.neighborUpdater = new InstantNeighborUpdater(this);
+        } else {
+            this.neighborUpdater = new CollectingNeighborUpdater(this, j);
+        }
+        // Horizon end - instantBlockUpdaterReintroduced
         this.registryAccess = iregistrycustom;
         this.damageSources = new DamageSources(iregistrycustom);
         // CraftBukkit start
