From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Wed, 12 Jun 2024 16:33:00 +0400
Subject: [PATCH] Stick can change ArmorStand arm status

Powered by Leaves: https://github.com/LeavesMC/Leaves

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index 33702c249963ed46407587df6e4859f1a07e5ec7..2b853477e032712fde71a24258b2e87c86019ff1 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -115,4 +115,7 @@ public final class HorizonConfig {
     @GlobalConfig(name = "redstone-wire-dont-connect-if-on-trapdoor", category = {"features"})
     public static boolean redstoneDontCantOnTrapDoor = false;
 
+    @GlobalConfig(name = "stick-change-armorstand-arm-status", category = {"utils"})
+    public static boolean stickChangeArmorStandArmStatus = false;
+
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index c0383771b08e1d89dfa4996b979248da1de8aca1..c70bbb69f294d4c4a9f4e450da1e2ba1f0c73d2d 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -395,6 +395,12 @@ public class ArmorStand extends LivingEntity {
                         return InteractionResult.SUCCESS;
                     }
                 } else {
+                    // Horizon start - stick can change ArmorStand arm status
+                    if (dev.horizonmc.horizon.HorizonConfig.stickChangeArmorStandArmStatus && itemstack.is(Items.STICK) && player.isShiftKeyDown()) {
+                        setShowArms(!isShowArms());
+                    }
+                    // Horizon end - stick can change ArmorStand arm status
+
                     if (this.isDisabled(enumitemslot)) {
                         return InteractionResult.FAIL;
                     }
