From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Tue, 16 Apr 2024 20:39:52 +0400
Subject: [PATCH] Carpet Protocol

Powered by Leaves: https://github.com/LeavesMC/Leaves

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index 77e2a44bfce4dbad192a8a1ee93520ada0607eeb..b3a2272cac44f4b4645d412fd74a9f65f99427b4 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -204,4 +204,7 @@ public final class HorizonConfig {
     @GlobalConfig(name = "redstone-wire-dont-connect-if-on-trapdoor", category = {"features"})
     public static boolean redstoneDontCantOnTrapDoor = false;
 
+    @GlobalConfig(name = "horizon-carpet-support", category = {"protocols"})
+    public static boolean horizonCarpetSupport = false;
+
 }
\ No newline at end of file
diff --git a/src/main/java/dev/horizonmc/horizon/protocol/CarpetServerProtocol.java b/src/main/java/dev/horizonmc/horizon/protocol/CarpetServerProtocol.java
new file mode 100644
index 0000000000000000000000000000000000000000..bf1b41e3edaeddf9508bfa5a59739f33b08b107f
--- /dev/null
+++ b/src/main/java/dev/horizonmc/horizon/protocol/CarpetServerProtocol.java
@@ -0,0 +1,119 @@
+package dev.horizonmc.horizon.protocol;
+
+import dev.horizonmc.horizon.HorizonConfig;
+import dev.horizonmc.horizon.HorizonLogger;
+import dev.horizonmc.horizon.protocol.core.HorizonCustomPayload;
+import dev.horizonmc.horizon.protocol.core.HorizonProtocol;
+import dev.horizonmc.horizon.protocol.core.ProtocolHandler;
+import dev.horizonmc.horizon.protocol.core.ProtocolUtils;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+@HorizonProtocol(namespace = "carpet")
+public class CarpetServerProtocol {
+
+    public static final String PROTOCOL_ID = "carpet";
+    public static final String VERSION = "horizon-carpet-1.0.0";
+
+    private static final ResourceLocation HELLO_ID = CarpetServerProtocol.id("hello");
+
+    private static final String HI = "69";
+    private static final String HELLO = "420";
+
+    @Contract("_ -> new")
+    public static @NotNull ResourceLocation id(String path) {
+        return new ResourceLocation(PROTOCOL_ID, path);
+    }
+
+    @ProtocolHandler.PlayerJoin
+    public static void onPlayerJoin(ServerPlayer player) {
+        if (HorizonConfig.horizonCarpetSupport) {
+            CompoundTag data = new CompoundTag();
+            data.putString(HI, VERSION);
+            ProtocolUtils.sendPayloadPacket(player, new CarpetPayload(data));
+        }
+    }
+
+    @ProtocolHandler.PayloadReceiver(payload = CarpetPayload.class, payloadId = "hello")
+    private static void handleHello(@NotNull ServerPlayer player, @NotNull CarpetServerProtocol.CarpetPayload payload) {
+        if (HorizonConfig.horizonCarpetSupport) {
+            if (payload.nbt.contains(HELLO)) {
+                HorizonLogger.LOGGER.info("Player " + player.getScoreboardName() + " joined with carpet " + payload.nbt.getString(HELLO));
+                CompoundTag data = new CompoundTag();
+                CarpetRules.write(data);
+                ProtocolUtils.sendPayloadPacket(player, new CarpetPayload(data));
+            }
+        }
+    }
+
+    public static class CarpetRules {
+
+        private static final Map<String, CarpetRule> rules = new HashMap<>();
+
+        public static void write(@NotNull CompoundTag tag) {
+            CompoundTag rulesNbt = new CompoundTag();
+            rules.values().forEach(rule -> rule.writeNBT(rulesNbt));
+
+            tag.put("Rules", rulesNbt);
+        }
+
+        public static void register(CarpetRule rule) {
+            rules.put(rule.name, rule);
+        }
+    }
+
+    public record CarpetRule(String identifier, String name, String value) {
+
+        @NotNull
+        @Contract("_, _, _ -> new")
+        public static CarpetRule of(String identifier, String name, Enum<?> value) {
+            return new CarpetRule(identifier, name, value.name().toLowerCase(Locale.ROOT));
+        }
+
+        @NotNull
+        @Contract("_, _, _ -> new")
+        public static CarpetRule of(String identifier, String name, boolean value) {
+            return new CarpetRule(identifier, name, Boolean.toString(value));
+        }
+
+        public void writeNBT(@NotNull CompoundTag rules) {
+            CompoundTag rule = new CompoundTag();
+            String key = name;
+
+            while (rules.contains(key)) {
+                key = key + "2";
+            }
+
+            rule.putString("Value", value);
+            rule.putString("Manager", identifier);
+            rule.putString("Rule", name);
+            rules.put(key, rule);
+        }
+    }
+
+    public record CarpetPayload(CompoundTag nbt) implements HorizonCustomPayload<CarpetPayload> {
+
+        @New
+        public CarpetPayload(ResourceLocation location, FriendlyByteBuf buf) {
+            this(buf.readNbt());
+        }
+
+        @Override
+        public void write(FriendlyByteBuf buf) {
+            buf.writeNbt(nbt);
+        }
+
+        @Override
+        @NotNull
+        public ResourceLocation id() {
+            return HELLO_ID;
+        }
+    }
+}
