From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Tue, 16 Apr 2024 16:08:08 +0400
Subject: [PATCH] Disable Packet Limit

Powered by Leaves: https://github.com/LeavesMC/Leaves

diff --git a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
index dd051aca08ce2dc6a6b94ad5389e99c317b9eeeb..331d609f84bc2597e7eb328d8dc5b9904c6225a9 100644
--- a/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
+++ b/src/main/java/dev/horizonmc/horizon/HorizonConfig.java
@@ -130,4 +130,7 @@ public final class HorizonConfig {
     @GlobalConfig(name = "appleskin-protocol", category = {"protocols"})
     public static boolean appleskinProtocol = false;
 
+    @GlobalConfig(name = "disable-packet-limit", category = {"features"})
+    public static boolean disablePacketLimit = true;
+
 }
\ No newline at end of file
diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 0a3f2f5b474910af3cabf3bab5310c0e970aed68..7b850102ba2db68e33fa43eacdd0a25916a012f8 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -267,7 +267,7 @@ public class GlobalConfiguration extends ConfigurationPart {
             }
 
             public boolean isEnabled() {
-                return this.interval > 0.0 && this.maxPacketRate > 0.0;
+                return !dev.horizonmc.horizon.HorizonConfig.disablePacketLimit && (this.interval > 0.0 && this.maxPacketRate > 0.0); // Horizon - disable packet limit
             }
 
             public enum ViolateAction {
