From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Sat, 27 Apr 2024 10:15:23 +0400
Subject: [PATCH] Fix Annotations


diff --git a/src/main/java/dev/horizonmc/horizon/entity/CraftBotManager.java b/src/main/java/dev/horizonmc/horizon/entity/CraftBotManager.java
index 62085e6f875d80711b590a70044a276f2798fc0e..f1c82952d1b9ef775e6366a8efb456fa4ad68b92 100644
--- a/src/main/java/dev/horizonmc/horizon/entity/CraftBotManager.java
+++ b/src/main/java/dev/horizonmc/horizon/entity/CraftBotManager.java
@@ -82,12 +82,12 @@ public class CraftBotManager implements BotManager {
     }
 
     @Override
-    public boolean registerCustomBotAction(String name, CustomBotAction action) {
+    public boolean registerCustomBotAction(@NotNull String name, @NotNull CustomBotAction action) {
         return Actions.register(new CraftCustomBotAction(name, action));
     }
 
     @Override
-    public boolean unregisterCustomBotAction(String name) {
+    public boolean unregisterCustomBotAction(@NotNull String name) {
         return Actions.unregister(name);
     }
 }
