From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GideonWhite1029 <gideonwhite1029@gmail.com>
Date: Fri, 12 Jul 2024 15:38:01 +0400
Subject: [PATCH] Matter Seed Command


diff --git a/src/main/java/net/minecraft/server/commands/SeedCommand.java b/src/main/java/net/minecraft/server/commands/SeedCommand.java
index 0b500b19a99fa6c2740c0db350a166462668df9c..1b9bb670a22bad222a32aa159dd1a91698e86b5c 100644
--- a/src/main/java/net/minecraft/server/commands/SeedCommand.java
+++ b/src/main/java/net/minecraft/server/commands/SeedCommand.java
@@ -6,12 +6,25 @@ import net.minecraft.commands.Commands;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
 
+// Matter start
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.HoverEvent;
+import su.plo.matter.Globals;
+// Matter end
+
 public class SeedCommand {
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher, boolean dedicated) {
         dispatcher.register(Commands.literal("seed").requires(source -> !dedicated || source.hasPermission(2)).executes(context -> {
             long l = context.getSource().getLevel().getSeed();
+            // Matter start - Seed Command
+            Globals.setupGlobals(context.getSource().getLevel());
+            String seedStr = Globals.seedToString(Globals.worldSeed);
             Component component = ComponentUtils.copyOnClickText(String.valueOf(l));
+            Component featureSeedComponent = ComponentUtils.copyOnClickText(seedStr);
             context.getSource().sendSuccess(() -> Component.translatable("commands.seed.success", component), false);
+            context.getSource().sendSuccess(() -> Component.translatable("Feature seed: %s", featureSeedComponent), false);
+            // Matter end
             return (int)l;
         }));
     }
